### PR TITLE
settings: Fix check for duplicate when new value is smaller

### LINKED ISSUE #14840: settings: settings_save_one() doesn't always seem to store data, even if it returns success
I have the following configuration:

 - tests/bluetooth/shell
 - BOARD=nrf52840_mdk

I was playing around with adding and removing identities, and noticed that deleting identities didn't work. On the settings level this maps to the keys bt/id and bt/irk. First one is multiples of seven bytes (sizeof(bt_addr_le_t)) and the second is multiples of 16 bytes (sizeof(IRK)).

What it looks like is that if the new value for key is bigger than the previous value then it gets stored and later recovered (through settings_load) just fine, however if the new value is smaller than the previous (e.g. the identity deletion case) then the new value is not stored. I also noticed that I could work around this by inserting an "empty" entry before the next one, e.g. calling settings_save_one(NULL, 0) before the actual value.

### PR DESCRIPTION
**Pull request by jhedberg**
The code for checking duplicates in the existing settings store was
incorrectly identifying a new value with the same initial content but
shorter length as a duplicate. Fix this by doing an early check on a
new vs old length mismatch, and immediately flag this as not a
duplicate.

Fixes #14840

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

### CODE REVIEW

**Review by nvlsianpu — APPROVED** [review]
Thanks for the fix!

**Review by Laczen — APPROVED** [review]

**Comment by jhedberg** [comment]
Question: is it valid that you have `val == NULL && len != 0` or `val != NULL && len == 0`? The code seems to have made the assumption that it's possible, which creates quite complex logic.

**Comment by jhedberg** [comment]
@Laczen @nvlsianpu 	I've updated the PR with a second patch which enforces `val == NULL && len == 0` and `val != NULL && len > 0`. This helps further simplify the original patch of this PR. Please re-review.

**Comment by Laczen** [comment]
> Question: is it valid that you have `val == NULL && len != 0` or `val != NULL && len == 0`? The code seems to have made the assumption that it's possible, which creates quite complex logic.

Yes, these are all valid, to avoid any complex logic I map 'val == NULL || LEN == 0' to a delete boolean in the reworked settings module. It would be better however to use the settings_delete(name) if you need to delete a entry.

File: subsys/settings/src/settings_store.c [hunk 1]
```diff
@@ -155,6 +152,14 @@ int settings_save_one(const char *name, void *value, size_t val_len)
 		return -ENOENT;
 	}
 
+	if (val_len > 0 && value == NULL) {
+		return -EINVAL;
+	}
+
+	if (val_len == 0) {
```
Comment (nvlsianpu) [inline subsys/settings/src/settings_store.c]: It is not required as all checks are done using lengths compassion.

Comment (jhedberg) [reply subsys/settings/src/settings_store.c]: > Correction is not required as all checks are done using lengths comparasion.

@nvlsianpu good point. PR has been updated - now the first patch only adds the -EINVAL return.

**Comment by jhedberg** [comment]
@Laczen @nvlsianpu are you ok with the latest version of this PR? All the checks have passed, and it has the two required reviews, so I'd be able to merge it.

**Comment by Laczen** [comment]
LGTM

**Comment by nvlsianpu** [comment]
LGTM as well.

**Pull request merged**